File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    private final void backtrack(List<String> validParens, StringBuilder paren, int n, int open, int close) {
        if (open == n && close == n) {
            validParens.add(paren.toString());
            return ;
        }

        if (open < n) {
            paren.append('(');
            backtrack(validParens, paren, n, open + 1, close);
            paren.delete(paren.length() - 1, paren.length());
        }

        if (close < open) {
            paren.append(')');
            backtrack(validParens, paren, n, open, close + 1);
            paren.delete(paren.length() - 1, paren.length());
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> validParens = new ArrayList<>();
        StringBuilder paren = new StringBuilder();
        backtrack(validParens, paren, n, 0, 0);
        return validParens;
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n, string paren = "", int open = 0, int close = 0) {
        vector<string> validParens;
        if (open == n && close == n) {
            validParens.push_back(paren);
            return validParens;
        }

        if (open < n) {
            vector<string> temp = generateParenthesis(n, paren + "(", open + 1, close);
            for (string& t: temp) {
                validParens.push_back(t);
            }
        }

        if (close < open) {
            vector<string> temp = generateParenthesis(n, paren + ")", open, close + 1);
            for (string& t: temp) {
                validParens.push_back(t);
            }
        }

        return validParens;
    }
};